src/server/game/Entities/Player/Player.cpp

```diff
@@ -15287,10 +15287,7 @@ void Player::RewardQuest(Quest const* quest, uint32 reward, Object* questGiver,
     if (quest->GetRewSpell() > 0)
     {
         SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(quest->GetRewSpell());
-        if (questGiver && questGiver->isType(TYPEMASK_UNIT) &&
-            !spellInfo->HasEffect(DIFFICULTY_NONE, SPELL_EFFECT_LEARN_SPELL) &&
-            !spellInfo->HasEffect(DIFFICULTY_NONE, SPELL_EFFECT_CREATE_ITEM) &&
-            !spellInfo->HasEffect(DIFFICULTY_NONE, SPELL_EFFECT_APPLY_AURA))
+        if (questGiver && questGiver->isType(TYPEMASK_UNIT) && spellInfo->HasTargetType(TARGET_UNIT_TARGET_ANY))
```

Comment:
TARGET_UNIT_TARGET_ANY or TARGET_UNIT_TARGET_ALLY should be casted by questgiver. Rest seems fine.
Ref https://gist.github.com/mdX7/683cc9453a399fc8a93e2a2bdfab3b73